File: consumer/pact-jvm-consumer/src/main/scala/au/com/dius/pact/consumer/MockProvider.scala
package au.com.dius.pact.consumer

import au.com.dius.pact.core.model.{Pact => PactModel, _}
import au.com.dius.pact.model.{MockHttpsKeystoreProviderConfig, MockHttpsProviderConfig, MockProviderConfig}
import com.typesafe.scalalogging.StrictLogging

import scala.util.Try

trait MockProvider[I <: Interaction] {
  def config: MockProviderConfig
  def session: PactSession
  def start(pact: PactModel[I]): Unit
  def run[T](code: => T): Try[T]
  def runAndClose[T](pact: PactModel[I])(code: => T): Try[(T, PactSessionResults)]
  def stop(): Unit
}

object DefaultMockProvider {
  
  def withDefaultConfig(pactVersion: PactSpecVersion = PactSpecVersion.V3) =
    apply(MockProviderConfig.createDefault(pactVersion))
  
  // Constructor providing a default implementation of StatefulMockProvider.
  // Users should not explicitly be forced to choose a variety.
  def apply(config: MockProviderConfig): StatefulMockProvider[RequestResponseInteraction] =
    config match {
      case httpsConfig: MockHttpsProviderConfig => new UnfilteredHttpsMockProvider(httpsConfig)
      case httpsKeystoreConfig: MockHttpsKeystoreProviderConfig => new UnfilteredHttpsKeystoreMockProvider(httpsKeystoreConfig)
      case _ => new UnfilteredMockProvider(config)
    }
}

// TODO: eliminate horrid state mutation and synchronisation.  Reactive stuff to the rescue?
abstract class StatefulMockProvider[I <: Interaction] extends MockProvider[I] with StrictLogging {
  private var sessionVar = PactSession.empty
  private var pactVar: Option[PactModel[I]] = None

  private def waitForRequestsToFinish() = Thread.sleep(100)

  def session: PactSession  = sessionVar
  def pact: Option[PactModel[I]] = pactVar
  
  def start(): Unit
  
  override def start(pact: PactModel[I]): Unit = synchronized {
    pactVar = Some(pact)
    sessionVar = PactSession.forPact(pact)
    start()
  }

  override def run[T](code: => T): Try[T] = {
    Try {
      val codeResult = code
      waitForRequestsToFinish()
      codeResult
    }
  }

  override def runAndClose[T](pact: PactModel[I])(code: => T): Try[(T, PactSessionResults)] = {
    Try {
      try {
        start(pact)
        val codeResult = code
        waitForRequestsToFinish()
        (codeResult, session.remainingResults)
      } finally {
        stop()
      }
    }
  }

  final def handleRequest(req: Request): Response = synchronized {
    logger.debug("Received request: " + req)
    val (response, newSession) = session.receiveRequest(req)
    logger.debug("Generating response: " + response)
    sessionVar = newSession
    response
  }
}

File: pact-jvm-server/src/main/kotlin/au/com/dius/pact/server/MockProvider.kt
package au.com.dius.pact.server

import au.com.dius.pact.consumer.model.MockHttpsKeystoreProviderConfig
import au.com.dius.pact.consumer.model.MockHttpsProviderConfig
import au.com.dius.pact.consumer.model.MockProviderConfig
import au.com.dius.pact.core.model.IResponse
import au.com.dius.pact.core.model.Pact
import au.com.dius.pact.core.model.PactSpecVersion
import au.com.dius.pact.core.model.Request
import au.com.dius.pact.core.support.handleWith
import au.com.dius.pact.core.support.Result
import io.github.oshai.kotlinlogging.KotlinLogging

private val logger = KotlinLogging.logger {}

interface MockProvider {
  val config: MockProviderConfig
  val session: PactSession
  fun start(pact: Pact)
  fun <T> run(code: () -> T): Result<T, Exception>
  fun <T> runAndClose(pact: Pact, code: () -> T): Result<Pair<T, PactSessionResults>, Exception>
  fun stop()
}

object DefaultMockProvider {

  fun withDefaultConfig(pactVersion: PactSpecVersion = PactSpecVersion.V3) =
    apply(MockProviderConfig.createDefault(pactVersion))

  // Constructor providing a default implementation of StatefulMockProvider.
  // Users should not explicitly be forced to choose a variety.
  fun apply(config: MockProviderConfig): StatefulMockProvider =
    when (config) {
      is MockHttpsProviderConfig -> KTorHttpsKeystoreMockProvider(config)
//      is MockHttpsProviderConfig -> UnfilteredHttpsMockProvider(config)
      else -> KTorMockProvider(config)
    }
}

abstract class StatefulMockProvider: MockProvider {
  private var sessionVar = PactSession.empty
  private var pactVar: Pact? = null

  private fun waitForRequestsToFinish() = Thread.sleep(100)

  override val session: PactSession
    get() = sessionVar
  val pact: Pact?
    get() = pactVar

  abstract fun start()

  @Synchronized
  override fun start(pact: Pact) {
    pactVar = pact
    sessionVar = PactSession.forPact(pact)
    start()
  }

  override fun <T> run(code: () -> T): Result<T, Exception> {
    return handleWith {
      val codeResult = code()
      waitForRequestsToFinish()
      codeResult
    }
  }

  override fun <T> runAndClose(pact: Pact, code: () -> T): Result<Pair<T, PactSessionResults>, Exception> {
    return handleWith {
      try {
        start(pact)
        val codeResult = code()
        waitForRequestsToFinish()
        (codeResult to session.remainingResults())
      } finally {
        stop()
      }
    }
  }

  @Synchronized
  fun handleRequest(req: Request): IResponse {
    logger.debug { "Received request: $req" }
    val (response, newSession) = session.receiveRequest(req)
    logger.debug { "Generating response: $response" }
    sessionVar = newSession
    return response
  }
}
